Gifski/Gifski.swift

```diff
@@ -1,78 +1,97 @@
 import Foundation
 import AVFoundation
 
-final class Gifski {
-	private var progress: Progress!
-	private var observation: NSKeyValueObservation?
-
-	// `progress.fractionCompleted` is KVO-compliant, but we expose this for convenience
-	var onProgress: ((_ progress: Progress) -> Void)?
-
-	static func convertFile(
-		_ inputUrl: URL,
-		outputUrl: URL,
-		quality: Double = 1,
-		dimensions: CGSize? = nil,
-		frameRate: Int? = nil
-		) -> Gifski {
-		let gifski = Gifski()
-		gifski.convertFile(
-			inputUrl,
-			outputUrl: outputUrl,
-			quality: quality,
-			dimensions: dimensions,
-			frameRate: frameRate
-		)
-		return gifski
+enum Result {
+	case success
+	case error(GifskiConversionError)
+}
+
+extension Result {
+	func forceSuccess() {
+		if case .error(let error) = self {
+			fatalError(error.localizedDescription)
+		}
 	}
+}
 
-	/**
-	- parameters:
-		- frameRate: Clamped to 5...30. Uses the frame rate of `inputUrl` if not specified.
-	*/
-	private func convertFile(
-		_ inputUrl: URL,
-		outputUrl: URL,
-		quality: Double = 1,
-		dimensions: CGSize? = nil,
-		frameRate: Int? = nil
-	) {
-		progress = Progress(parent: .current(), userInfo: [.fileURLKey: outputUrl])
-		progress.fileURL = outputUrl
-		progress.publish()
-
-		observation = progress.observe(\.fractionCompleted) { progress, _ in
-			DispatchQueue.main.async {
-				self.onProgress?(progress)
-			}
+enum GifskiConversionError: LocalizedError {
+	case invalidSettings
+	case generateFrameFailed
+	case addFrameFailed(GifskiWrapperError)
+	case endAddingFramesFailed(GifskiWrapperError)
+	case writeFailed(GifskiWrapperError)
+
+	var errorDescription: String? {
+		switch self {
+		case .invalidSettings:
+			return "Invalid settings"
+		case .generateFrameFailed:
+			return "Failed to generate frame"
+		case .addFrameFailed(let error):
+			return "Failed to add frame, with underlying error: \(error.localizedDescription)"
+		case .endAddingFramesFailed(let error):
+			return "Failed to end adding frames, with underlying error: \(error.localizedDescription)"
+		case .writeFailed(let error):
+			return "Failed to write to output, with underlying errror: \(error.localizedDescription)"
 		}
+	}
+}
 
+/**
+- parameters:
+- frameRate: Clamped to 5...30. Uses the frame rate of `inputUrl` if not specified.
+*/
+struct Conversion {
+	let input: URL
+	let output: URL
+	let quality: Double
+	let dimensions: CGSize
+	let frameRate: Int?
+
+	init(input: URL, output: URL, quality: Double = 1, dimensions: CGSize = .zero, frameRate: Int? = nil) {
+		self.input = input
+		self.output = output
+		self.quality = quality
+		self.dimensions = dimensions
+		self.frameRate = frameRate
+	}
+}
+
+final class Gifski {
+	static func run(_ conversion: Conversion, completionHandler: ((Result) -> Void)?) {
```

Comment:
We are not passing anything back in the success case. Do we really need the Result type? Would it be better if the callback took an optional Error?

Reply:
Yeah, would probably be better. Although, we should keep cancelation in mind, which might require a Result object with a `cancelled` state.

Reply:
Removed the result type for now. Probably better to add in later when needed..